_posts/2018-6-11-Interpreter-Optimizations.md

```diff
@@ -0,0 +1,209 @@
+built. The BEAM interpreter is built using generated C89 code. Dispatching
+is done using [direct threading] which leverages the GCC extension [labels as values].
+It is the job of the [beam\_makeops] perl script to take the input files
+and create the C files.
+
+There is also a set of transformation rules generated that is used by the
+[transform\_engine] in the beam code loader to do several peephole optimization.
+The optimization include, but are not limited to, instruction combining,
+instruction specialization and dead code elimination.
+
+There are three seperate types of input files used.
+
+* [genops.tab] contains a listing of the instructions that the compiler emits.
+* [ops.tab] contains load time transformations done to the code.
+* [instrs.tab] contain the implementation of each instruction.
+
+There is a description of the syntax and semantics of the different files
+in the [internal beam\_makeops documentation]. The largest difference
+inbetween the new and old way of doing the code generation is that
+in OTP 21 _all_ the instructions are now generated, instead of about 75%.
+This allowed us to do architecture specific optimizations for all instructions
+when generating the code for the instructions.
+
+[interpreter]: https://en.wikipedia.org/wiki/Interpreter_(computing)
+[direct threading]: https://en.wikipedia.org/wiki/Threaded_code#Direct_threading
+[labels as values]: https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html
+[beam\_makeops]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/utils/beam_makeops
+[genops.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/lib/compiler/src/genops.tab
+[ops.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/ops.tab
+[instrs.tab]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/instrs.tab
+[internal beam\_makeops documentation]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/internal_doc/beam_makeops.md
+[transform\_engine]: https://github.com/erlang/otp/blob/OTP-21.0-rc1/erts/emulator/beam/beam_load.c#L5213
+
+## Packing the dispatch target address ##
+
+On 64-bit machines all pointers are 8 byte large, this is also the case for
+the pointer that you get when taking the address of a label. So for a small
+instruction such as move_cx (move a constant to a x register), 3 words of
+memory is needed.
+
+         +--------------------+--------------------+
+    I -> |                            &&lb_move_cx |
+         +--------------------+--------------------+
+         |                                      40 |
+         +--------------------+--------------------+
+         |                        Tagged atom 'id' |
+         +--------------------+--------------------+
+
+One word for the instruction, one word for the literal and then the target x
+register actually only needs 2 bytes, but since all code is word aligned it
+gets its own word.
+
```

Comment:
Change order to match ascii-art: Intruction word, target x register and then tagged literal.